File: Scr_Java/modelo/Cozy.java
package modelo;

// Enemigo del juego. Avanza por la ruta del mapa segun su velocidad
// y se retira cuando muere o llega al final (ver ColaEnemigos.actualizarQuantum).
public class Cozy {
    private final String tipo;
    private int pv;
    private final int pvMax;
    private int indiceRuta;
    private final int velocidad;
    private final int recompensa;

    public Cozy(String tipo, int pvMax, int velocidad, int recompensa) {
        this.tipo = tipo;
        this.pvMax = pvMax;
        this.pv = pvMax;
        this.velocidad = velocidad;
        this.recompensa = recompensa;
        this.indiceRuta = 0;
    }

    // Nunca deja el pv negativo, asi estaVivo() queda bien definido con pv > 0.
    public void recibirDano(int dano) {
        pv = Math.max(0, pv - dano);
    }

    public boolean estaVivo() {
        return pv > 0;
    }

    public void avanzar() {
        indiceRuta += velocidad;
    }

    // "-1" porque el ultimo indice valido de la ruta es longitudRuta - 1
    // (la fuente de energia).
    public boolean llegoAlFinal(int longitudRuta) {
        return indiceRuta >= longitudRuta - 1;
    }

    public String getTipo() {
        return tipo;
    }

    public int getPv() {
        return pv;
    }

    public int getPvMax() {
        return pvMax;
    }

    public int getIndiceRuta() {
        return indiceRuta;
    }

    public int getVelocidad() {
        return velocidad;
    }

    public int getRecompensa() {
        return recompensa;
    }
}

File: cpp/modelo/Cozy.cpp
#include "Cozy.h"

#include <algorithm>

// Constructor que inicializa las caracteristicas principales del enemigo.
Cozy::Cozy(const std::string& tipo, int pvMax, int velocidad, int recompensa)
    : tipo(tipo),
      pv(pvMax),
      pvMax(pvMax),
      indiceRuta(0),
      velocidad(velocidad),
      recompensa(recompensa) {
}

// Nunca deja el pv negativo, asi estaVivo() queda bien definido con pv > 0.
void Cozy::recibirDano(int dano) {
    pv = std::max(0, pv - dano);
}

// Determina si el enemigo continua con puntos de vida.
bool Cozy::estaVivo() const {
    return pv > 0;
}

// Incrementa la posicion del enemigo dentro de la ruta
// de acuerdo con su velocidad.
void Cozy::avanzar() {
    indiceRuta += velocidad;
}

// "-1" porque el ultimo indice valido de la ruta es longitudRuta - 1
// (la fuente de energia).
bool Cozy::llegoAlFinal(int longitudRuta) const {
    return indiceRuta >= longitudRuta - 1;
}

// Retorna el tipo de enemigo.
const std::string& Cozy::getTipo() const {
    return tipo;
}

// Retorna los puntos de vida actuales del enemigo.
int Cozy::getPv() const {
    return pv;
}

// Retorna los puntos de vida maximos del enemigo.
int Cozy::getPvMax() const {
    return pvMax;
}

// Retorna el indice actual del enemigo dentro de la ruta.
int Cozy::getIndiceRuta() const {
    return indiceRuta;
}

// Retorna la velocidad de movimiento del enemigo.
int Cozy::getVelocidad() const {
    return velocidad;
}

// Retorna la recompensa otorgada al eliminar al enemigo.
int Cozy::getRecompensa() const {
    return recompensa;
}